File: backend/formulas.py
"""Body composition formulas — fiéis às referências originais.

Todas as fórmulas recebem dicionário `p` com pregas (mm) e outros dados.
Devolvem valores em % ou kg. Se dados insuficientes -> None.
"""
from math import log10


def sum_pregas_7(p: dict) -> float | None:
    """Somatório 7 pregas: peito, tricipital, subescapular, axilar,
    suprailiaca, abdominal, coxa."""
    keys = ["peito", "tricipital", "subescapular", "axilar",
            "suprailiaca", "abdominal", "coxa"]
    vals = [p.get(k) for k in keys]
    if any(v is None for v in vals):
        return None
    return float(sum(vals))


def sum_pregas_8(p: dict) -> float | None:
    """Somatório 8 pregas: adiciona 'supraespinhal' ao 7-pregas."""
    keys = ["peito", "tricipital", "subescapular", "axilar",
            "suprailiaca", "abdominal", "coxa", "supraespinhal"]
    vals = [p.get(k) for k in keys]
    if any(v is None for v in vals):
        return None
    return float(sum(vals))


def bmi(weight_kg: float | None, height_cm: float | None) -> float | None:
    if not weight_kg or not height_cm:
        return None
    h = height_cm / 100.0
    if h <= 0:
        return None
    return round(weight_kg / (h * h), 2)


# ---------- Jackson-Pollock (7 skinfolds) ----------
def bf_jackson_pollock(p: dict, sex: str, age: float) -> float | None:
    """Jackson & Pollock 7 pregas — % MG.
    Densidade -> Siri para % MG.
    """
    s = sum_pregas_7(p)
    if s is None or age is None:
        return None
    if sex == "M":
        density = 1.112 - 0.00043499 * s + 0.00000055 * (s * s) - 0.00028826 * age
    else:
        density = 1.097 - 0.00046971 * s + 0.00000056 * (s * s) - 0.00012828 * age
    if density <= 0:
        return None
    bf = (495.0 / density) - 450.0
    return round(bf, 2)


# ---------- Withers (males 7 pregas) ----------
def bf_withers(p: dict, sex: str) -> float | None:
    """Withers et al. 1987 — 7 pregas (atletas).
    Homens: D = 1.0988 - 0.0004 * S7
    Mulheres: D = 1.17484 - 0.07229 * log10(S7)
    """
    s = sum_pregas_7(p)
    if s is None or s <= 0:
        return None
    if sex == "M":
        density = 1.0988 - 0.0004 * s
    else:
        density = 1.17484 - 0.07229 * log10(s)
    if density <= 0:
        return None
    bf = (495.0 / density) - 450.0
    return round(bf, 2)


# ---------- Reilly & Wallace (soccer players) ----------
def bf_reilly_wallace(p: dict) -> float | None:
    """Reilly et al. 2009 — jogadores de futebol.
    %MG = 5.174 + 0.124*(abdominal) + 0.147*(coxa) + 0.196*(tricipital)
          + 0.130*(gemeo)
    (mm de pregas)
    """
    keys = ["abdominal", "coxa", "tricipital", "gemeo"]
    vals = {k: p.get(k) for k in keys}
    if any(v is None for v in vals.values()):
        return None
    bf = (5.174
          + 0.124 * vals["abdominal"]
          + 0.147 * vals["coxa"]
          + 0.196 * vals["tricipital"]
          + 0.130 * vals["gemeo"])
    return round(bf, 2)


# ---------- Evans 2005 (athletes 3-site) ----------
def bf_evans(p: dict, sex: str, ethnicity: str = "caucasiano") -> float | None:
    """Evans et al. 2005 — atletas.
    Homens: %MG = 8.997 + 0.24658*(S3) - 6.343*(1 se afro) - 1.998*(1 se branco)
    Onde S3 = abdominal + coxa + tricipital
    Simplificação usada no JSX: coefs por etnia.
    """
    s = 0.0
    for k in ("abdominal", "coxa", "tricipital"):
        v = p.get(k)
        if v is None:
            return None
        s += v
    is_afro = (ethnicity or "").lower().startswith("afr")
    if sex == "M":
        bf = 8.997 + 0.24658 * s - (6.343 if is_afro else 0.0) - (1.998 if not is_afro else 0.0)
    else:
        # Fórmula geral para mulheres (Evans 2005)
        bf = 10.566 + 0.20345 * s - (6.343 if is_afro else 0.0) - (1.998 if not is_afro else 0.0)
    return round(bf, 2)


# ---------- Lee et al. 2000 — Massa Muscular (SMM em kg) ----------
def muscle_mass_lee(height_cm: float | None,
                    perimeters: dict,
                    sex: str,
                    age: float | None,
                    ethnicity: str = "caucasiano") -> float | None:
    """Lee et al. 2000 — Skeletal Muscle Mass (kg).
    SMM = Ht * (0.00744*CAG^2 + 0.00088*CTG^2 + 0.00441*CCG^2)
          + 2.4*sex - 0.048*age + race + 7.8
    onde:
        CAG = perimetro braço corrigido (cm - preg triciptal em cm)
        CTG = perimetro coxa corrigido (cm - preg coxa em cm)
        CCG = perimetro gémeo corrigido (cm - preg gémeo em cm)
    sex = 1 se M, 0 se F. race = 1.1 (asiático), 1.4 (afro), 0 (caucas.)
    Neste app usamos perímetros medidos (não corrigidos) por simplicidade
    porque não pedimos pregas nos mesmos sítios em cm.
    """
    if not height_cm or age is None:
        return None
    ht = height_cm / 100.0
    cag = perimeters.get("braco")
    coxa_perim = perimeters.get("coxaD") or perimeters.get("coxaE")
    ccg = perimeters.get("gemeo")
    if cag is None or coxa_perim is None or ccg is None:
        return None
    sex_v = 1.0 if sex == "M" else 0.0
    eth = (ethnicity or "").lower()
    if eth.startswith("asi"):
        race = 1.1
    elif eth.startswith("afr"):
        race = 1.4
    else:
        race = 0.0
    smm = (ht * (0.00744 * cag * cag
                 + 0.00088 * coxa_perim * coxa_perim
                 + 0.00441 * ccg * ccg)
           + 2.4 * sex_v - 0.048 * age + race + 7.8)
    return round(smm, 2)


def status_from_bf(bf: float | None, sex: str) -> str:
    """Retorna faixa: Ótimo / Atenção / Alto conforme sexo."""
    if bf is None:
        return "—"
    if sex == "M":
        if bf < 12:
            return "Ótimo"
        if bf < 18:
            return "Atenção"
        return "Alto"
    # F
    if bf < 20:
        return "Ótimo"
    if bf < 26:
        return "Atenção"
    return "Alto"


def compute_all(evaluation: dict, athlete: dict) -> dict:
    """Recebe evaluation (pregas+perimetros+peso+age_at_eval) e athlete
    (sexo, etnia, altura_cm). Devolve dicionário com todos os cálculos."""
    p = evaluation.get("pregas", {}) or {}
    per = evaluation.get("perimetros", {}) or {}
    sex = athlete.get("sexo", "M")
    ethnicity = athlete.get("etnia", "caucasiano")
    height_cm = athlete.get("altura_cm")
    age = evaluation.get("age_at_eval") or athlete.get("idade")
    weight = evaluation.get("peso_kg")

    reilly = bf_reilly_wallace(p)
    evans = bf_evans(p, sex, ethnicity)
    jp = bf_jackson_pollock(p, sex, age) if age else None
    withers = bf_withers(p, sex)

    bfs = [x for x in (reilly, evans, jp, withers) if x is not None]
    bf_avg = round(sum(bfs) / len(bfs), 2) if bfs else None

    smm = muscle_mass_lee(height_cm, per, sex, age, ethnicity)
    imc = bmi(weight, height_cm)
    s7 = sum_pregas_7(p)
    s8 = sum_pregas_8(p)

    mg_kg = round(weight * bf_avg / 100.0, 2) if (weight and bf_avg is not None) else None
    ratio = round(smm / mg_kg, 2) if (smm and mg_kg and mg_kg > 0) else None

    return {
        "bf_reilly_wallace": reilly,
        "bf_evans": evans,
        "bf_jackson_pollock": jp,
        "bf_withers": withers,
        "bf_average": bf_avg,
        "muscle_mass_kg": smm,
        "fat_mass_kg": mg_kg,
        "mm_mg_ratio": ratio,
        "imc": imc,
        "sum_pregas_7": s7,
        "sum_pregas_8": s8,
        "status": status_from_bf(bf_avg, sex),
    }

File: frontend/src/lib/formulas.js
// Mirror EXATO do ficheiro Trofense_APP.jsx original para cálculo em tempo real
const PI_EXCEL = 3.1415;

const num = (p, ...keys) => {
  const out = [];
  for (const k of keys) {
    const v = p?.[k];
    if (v == null || v === "") return null;
    out.push(Number(v));
  }
  return out;
};

export function sumPregas8(p) {
  const v = num(p, "peito", "tricipital", "subescapular", "axilar", "suprailiaca", "abdominal", "coxa", "gemeo");
  return v ? +v.reduce((a, b) => a + b, 0).toFixed(2) : null;
}
export function sumPregas7(p) {
  const v = num(p, "peito", "tricipital", "subescapular", "axilar", "suprailiaca", "abdominal", "coxa");
  return v ? +v.reduce((a, b) => a + b, 0).toFixed(2) : null;
}

export function bmi(w, hCm) {
  if (!w || !hCm) return null;
  const h = hCm / 100;
  return +(w / (h * h)).toFixed(2);
}

export function bfReillyWallace(p) {
  const v = num(p, "coxa", "abdominal", "gemeo");
  if (!v) return null;
  const [coxa, abdom, gemeo] = v;
  return +(5.174 + 0.124 * coxa + 0.147 * abdom + 0.13 * gemeo).toFixed(2);
}

export function bfEvans7(p, sexNum, ethNum) {
  const v = num(p, "peito", "tricipital", "bicipital", "axilar", "suprailiaca", "abdominal", "coxa");
  if (!v) return null;
  const s = v.reduce((a, b) => a + b, 0);
  return +(10.566 + 0.12077 * s - 8.057 * sexNum - 2.545 * ethNum).toFixed(2);
}

export function bfEvans3(p, sexNum, ethNum) {
  const v = num(p, "abdominal", "coxa", "tricipital");
  if (!v) return null;
  const s = v.reduce((a, b) => a + b, 0);
  return +(8.997 + 0.24658 * s - 6.343 * sexNum - 1.998 * ethNum).toFixed(2);
}

export function bfJacksonPollock(p, age) {
  const v = num(p, "peito", "tricipital", "bicipital", "subescapular", "abdominal", "coxa", "suprailiaca");
  if (!v || !age) return null;
  const s = v.reduce((a, b) => a + b, 0);
  const d = 1.112 - 0.00043499 * s + 0.00000055 * s * s - 0.00028826 * age;
  if (d <= 0) return null;
  return +((4.95 / d - 4.5) * 100).toFixed(2);
}

export function bfWithers(p) {
  const v = num(p, "tricipital", "bicipital", "subescapular", "abdominal", "supraespinhal", "coxa", "gemeo");
  if (!v) return null;
  const s = v.reduce((a, b) => a + b, 0);
  const d = 1.0988 - 0.0004 * s;
  if (d <= 0) return null;
  return +(495 / d - 450).toFixed(2);
}

export function muscleMassLee(alturaCm, p, per, sexNum, ethNum, age) {
  if (!alturaCm || age == null) return null;
  const braco = per?.braco, coxaD = per?.coxaD ?? per?.coxaE, gemPer = per?.gemeo;
  const tri = p?.tricipital, sup = p?.supraespinhal, gemPreg = p?.gemeo;
  for (const v of [braco, coxaD, gemPer, tri, sup, gemPreg]) if (v == null || v === "") return null;
  const alturaM = alturaCm / 100;
  const mm =
    alturaM *
      (0.00744 * Math.pow(braco - (PI_EXCEL * tri) / 10, 2) +
        0.00088 * Math.pow(coxaD - (PI_EXCEL * sup) / 10, 2) +
        0.00441 * Math.pow(gemPer - (PI_EXCEL * gemPreg) / 10, 2)) +
    2.4 * sexNum -
    0.048 * age +
    ethNum +
    7.8;
  return +mm.toFixed(2);
}

// ---------- Bandas (pontos de corte do original) ----------
export function rwBand(rw) {
  if (rw == null) return { label: "—", color: "muted" };
  if (rw < 9) return { label: "Ótimo", color: "otimo" };
  if (rw <= 10) return { label: "Atenção", color: "atencao" };
  return { label: "Alto", color: "alto" };
}
export function soma8Band(s8) {
  if (s8 == null) return { label: "—", color: "muted" };
  if (s8 < 65) return { label: "Baixo", color: "otimo" };
  if (s8 <= 75) return { label: "Médio", color: "atencao" };
  return { label: "Alto", color: "alto" };
}
export function percMMBand(pm) {
  if (pm == null) return { label: "—", color: "muted" };
  if (pm >= 45) return { label: "Ótimo", color: "otimo" };
  if (pm >= 40) return { label: "Atenção", color: "atencao" };
  return { label: "Baixo", color: "alto" };
}
export function mmMgBand(r, sexNum) {
  if (r == null) return { label: "—", color: "muted" };
  const min = sexNum === 1 ? 4 : 2;
  return r >= min ? { label: "Normal", color: "otimo" } : { label: "Baixo", color: "alto" };
}
export function imcBand(imc) {
  if (imc == null) return { label: "—", color: "muted" };
  if (imc < 18.5) return { label: "Baixo peso", color: "alto" };
  if (imc < 25) return { label: "Normal", color: "otimo" };
  if (imc < 30) return { label: "Sobrepeso", color: "atencao" };
  return { label: "Obesidade", color: "alto" };
}

export function computeAll(evaluation, athlete) {
  const p = evaluation?.pregas || {};
  const per = evaluation?.perimetros || {};
  const sexNum = athlete?.sexo === "M" ? 1 : 0;
  const eth = (athlete?.etnia || "caucasiano").toLowerCase();
  const ethNum = eth.startsWith("afr") ? 1 : 0;
  const alturaCm = athlete?.altura_cm;
  const age = evaluation?.age_at_eval ?? athlete?.idade;
  const w = evaluation?.peso_kg;

  const rw = bfReillyWallace(p);
  const evans7 = bfEvans7(p, sexNum, ethNum);
  const evans3 = bfEvans3(p, sexNum, ethNum);
  const jp7 = age ? bfJacksonPollock(p, age) : null;
  const withers = bfWithers(p);
  const bfs = [rw, evans7, evans3, jp7, withers].filter((v) => v != null);
  const bf_average = bfs.length ? +(bfs.reduce((a, b) => a + b, 0) / bfs.length).toFixed(2) : null;
  const mm = muscleMassLee(alturaCm, p, per, sexNum, ethNum, age);
  const imc = bmi(w, alturaCm);
  const s8 = sumPregas8(p);
  const s7 = sumPregas7(p);
  const mg_kg = w && rw != null ? +((w * rw) / 100).toFixed(2) : null;
  const lean = w && mg_kg != null ? +(w - mg_kg).toFixed(2) : null;
  const mm_mg = mm && mg_kg ? +(mm / mg_kg).toFixed(2) : null;
  const perc_mm = mm && w ? +((mm / w) * 100).toFixed(2) : null;

  return {
    rw, evans7, evans3, jp7, withers,
    bf_average, muscle_mass_kg: mm, perc_mm,
    fat_mass_kg: mg_kg, lean_mass_kg: lean,
    mm_mg_ratio: mm_mg, imc,
    soma7: s7, soma8: s8,
    // legacy aliases
    bf_reilly_wallace: rw, bf_evans: evans3, bf_jackson_pollock: jp7, bf_withers: withers,
    status: rwBand(rw).label,
  };
}
